File: 정올/Intermediate_Coder/동적계획법/배낭채우기1.py
n, w = map(int, input().split())
jewels = sorted([list(map(int, input().split())) for _ in range(n)])

d = [0] * 10001

for i in range(1, w+1):
    m = 0
    for weight, value in jewels:
        if i < weight:
            break
        m = max(m, d[i - weight] + value)
    d[i] = m

print(d[w])

File: 정올/Intermediate_Coder/동적계획법/배낭채우기1.cpp
#include <bits/stdc++.h>
using namespace std;

struct Jewel {
    int weight, value;
};

vector<Jewel> jewels;
int n, w, dp[10050];

int main() {
    ios_base::sync_with_stdio(false); cin.tie(NULL); cout.tie(NULL);

    cin>>n>>w;
    int a, b;
    for(int i = 1; i <= n; i++) {
        cin>>a>>b;
        jewels.push_back({a, b});
    }

    for(int i = 1; i <= w; i++) {
        int m = 0;
        for(auto jewel : jewels)
            if(i >= jewel.weight)
                m = max(m, dp[i - jewel.weight] + jewel.value);
        dp[i] = m;
    }

    cout<<dp[w];
}
